src/jit/compiler.cpp

```diff
@@ -516,64 +487,450 @@ var_types    Compiler::argOrReturnTypeForStruct(unsigned size, CORINFO_CLASS_HAN
+                useType = TYP_UNKNOWN;
+            }
+#else  // 32BIT
+            // structSize of 4 with IsHfa, we must have an HFA of one float
+            useType = TYP_FLOAT;
+#endif
+        }
+        else
+        {
+            BYTE gcPtr = 0;
+            // Check if this pointer-sized struct is wrapping a GC object
+            info.compCompHnd->getClassGClayout(clsHnd, &gcPtr);
+            useType = getJitGCType(gcPtr);
+        }
         break;
 
     default:
-#if FEATURE_MULTIREG_RET
-        if (forReturn)
+        useType = TYP_UNKNOWN;
+        break;
+    }
+
+    return useType;
+}
+
+//-----------------------------------------------------------------------------
+// getArgTypeForStruct: 
+//     Get the type that is used to pass values of the given struct type.
+//     If you already have retrieved the struct size then it should be passed
+//     as the third optional argument as it allows us to avoid an extra call
+//     to getClassSize(clsHnd)
+//
+// Arguments:
+//    clsHnd       - the handle for the struct type
+//    wbPassStruct - information about how the struct is passed
+//    structSize   - the size of the struct type, 
+//                   or 0 if we should to call getClassSize(clsHnd)
+//
+// Return Value:
+//    When *wbPassStruct is SPK_PrimitiveType this method's return value:
+//       is the primitive type used to pass the struct.
+//    When *wbPassStruct is SPK_ByReference this method's return value:
+//       is TYP_UNKNOWN and the struct type is passed by reference to a copy
+//    When *wbPassStruct is SPK_ByValue or SPK_ByValueAsHfa the return value:
+//       can be TYP_STRUCT and the struct type is passed using multiple registers.
+//       can be TYP_UNKNOWN and the struct type is passed by value (x86 and ARM32 only)
+//
+// Assumptions:
+//    The size must be the size of the given type.
+//    The given class handle must be for a value type (struct).
+//
+// Notes:
+//    About returning TYP_STRUCT:
+//        Whenever the return type is TYP_STRUCT it always means multiple registers
+//         are used to pass this struct.
+//    About HFA types:
+//        When the clsHnd is a one element HFA type we return the appropriate
+//         floating point primitive type and *wbPassStruct is SPK_PrimitiveType
+//        If there are two or more elements in the HFA type then the return value 
+//         is TYP_STRUCT and *wbPassStruct is SPK_ByValueAsHfa
+//
+var_types  Compiler::getArgTypeForStruct(CORINFO_CLASS_HANDLE clsHnd,
+                                         structPassingKind*   wbPassStruct,
+                                         unsigned             structSize /* = 0 */)
+{
+    var_types useType = TYP_UNKNOWN;
+    *wbPassStruct     = SPK_Unknown;   // We must change this before we can return
+
+    if (structSize == 0)
+    {
+        structSize = info.compCompHnd->getClassSize(clsHnd);
+    }
+    assert(structSize > 0);
+
```

Comment:
I wonder if we can do some kind of "assume" here that would cause targets with "if (structSize <= 0)" to eliminate those cases.